server/render/index.js

```diff
@@ -40,25 +41,64 @@ function bumpStat( group, name ) {
 	}
 }
 
+export function getCacheKey( context ) {
+	const pathname = context.pathname;
+
+	if ( isEmpty( context.query ) || isEmpty( context.cacheQueryKeys ) ) {
+		return pathname;
+	}
+
+	const cachedQueryParams = pick( context.query, context.cacheQueryKeys );
+
+	if ( isEmpty( cachedQueryParams ) ) {
+		return pathname;
+	}
+
+	return pathname + '?' + qs.stringify( cachedQueryParams, { sort: ( a, b ) => a.localCompare( b ) } );
+}
+
 /**
 * Render and cache supplied React element to a markup string.
 * Cache is keyed by stringified element by default.
 *
-* @param {object} element - React element to be rendered to html
+* @param {object} context - React element to be rendered to html
 * @param {string} key - (optional) custom key
 * @return {string} The rendered Layout
 */
-export function render( element, key = JSON.stringify( element ) ) {
+export function renderLayout( context ) {
 	try {
 		const startTime = Date.now();
+		const isIsomorphic = isSectionIsomorphic( context.store.getState() );
+		const key = isIsomorphic ? getCacheKey( context ) : JSON.stringify( context.layout );
+
 		debug( 'cache access for key', key );
 
 		let renderedLayout = markupCache.get( key );
 		if ( ! renderedLayout ) {
 			bumpStat( 'calypso-ssr', 'loggedout-design-cache-miss' );
 			debug( 'cache miss for key', key );
-			renderedLayout = ReactDomServer.renderToString( element );
+			renderedLayout = ReactDomServer.renderToString( context.layout );
 			markupCache.set( key, renderedLayout );
+
+			// save the state of the store with the same key as the layout
+			if ( context.store ) {
+				let reduxSubtrees = [ 'documentHead' ];
+
+				if ( isIsomorphic ) {
+					reduxSubtrees = reduxSubtrees.concat( [ 'ui', 'themes' ] );
+				}
+
+				// Send state to client
+				context.initialReduxState = pick( context.store.getState(), reduxSubtrees );
+				const serverState = reducer( context.initialReduxState, { type: SERIALIZE } );
+				stateCache.set( key, serverState );
+			}
+		} else {
+			const serializedServerState = stateCache.get( key );
+			if ( serializedServerState ) {
+				context.initialReduxState = reducer( serializedServerState, { type: DESERIALIZE } );
+				context.store = createReduxStore( context.initialReduxState );
```

Comment:
Why this is required?
I don't see anything using that further down the line...

Reply:
All those access the store though:
https://github.com/Automattic/wp-calypso/pull/19036/files#diff-d92d930a21711a960b300aeb231d9056R135

Reply:
Aha, I see, thank you!
It's not that optimal to initialize the store just for that, but lets merge and optimize later...